app/channel_share.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/store"
+)
+
+// SharedChannels
+
+func (a *App) SaveSharedChannel(sc *model.SharedChannel) (*model.SharedChannel, error) {
+	// Check channel is not already shared.
+	if _, err := a.GetSharedChannel(sc.ChannelId); err == nil {
+		return nil, errors.New("channel is already shared.")
+	}
+
+	// check that channel exists.
+	if _, errApp := a.GetChannel(sc.ChannelId); errApp != nil {
+		return nil, fmt.Errorf("cannot share this channel: %v", errApp)
+	}
+
+	return a.Srv().Store.Channel().SaveSharedChannel(sc)
+}
+
+func (a *App) GetSharedChannel(channelId string) (*model.SharedChannel, error) {
+	return a.Srv().Store.Channel().GetSharedChannel(channelId)
+}
+
+func (a *App) GetSharedChannels(page int, perPage int, opts store.SharedChannelFilterOpts) ([]*model.SharedChannel, error) {
+	return a.Srv().Store.Channel().GetSharedChannels(page*perPage, perPage, opts)
+}
+
+func (a *App) GetSharedChannelsCount(opts store.SharedChannelFilterOpts) (int64, error) {
+	return a.Srv().Store.Channel().GetSharedChannelsCount(opts)
+}
+
+func (a *App) UpdateSharedChannel(sc *model.SharedChannel) (*model.SharedChannel, error) {
+	return a.Srv().Store.Channel().UpdateSharedChannel(sc)
+}
+
+func (a *App) DeleteSharedChannel(channelId string) (bool, error) {
+	return a.Srv().Store.Channel().DeleteSharedChannel(channelId)
+}
+
+// SharedChannelRemotes
+func (a *App) SaveSharedChannelRemote(remote *model.SharedChannelRemote) (*model.SharedChannelRemote, error) {
+	// Check channel is shared.
+	if _, err := a.GetSharedChannel(remote.ChannelId); err != nil {
+		return nil, errors.New("channel is not shared.")
+	}
```

Comment:
99% of the time it will say, "channel not shared:  resource:shared channel id:d4rftrgfhsrfe4".  I don't think the extra fields add much value.

Reply:
I think in these cases the best approach would be to check if the error wraps `store.ErrNotFound` (using `errors.As()`) and if not wrap the returned error. Otherwise we lose all the information which might be rare but still valuable in case of debugging.